File: framework/java/src/main/java/soda/unittest/validate/XMap.java
package soda.unittest.validate;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class XMap<K, V> {

    public static class XEntry<K, V> {
        public final K key;
        public V value;
        public XEntry(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }

    private ObjectFeature<K> feat;

    private Map<Long, List<XEntry<K,V>>> dict = new HashMap<>();

    private int size;

    public XMap(ObjectFeature<K> feat) {
        this.feat = feat;
    }

    private long _hash(K key) {
        return feat.hash(key);
    }

    public boolean contains(K key) {
        var e = _entry(key);
        return e != null;
    }

    public V get(K key) {
        var e = _entry(key);
        return e != null ? e.value : null;
    }

    public V getOrDefault(K key, V defaultValue) {
        V v = get(key);
        return v != null ? v : defaultValue;
    }

    private XEntry<K, V> _entry(K key) {
        long h = _hash(key);
        if (dict.containsKey(h)) {
            for (var entry : dict.get(h)) {
                if (feat.isEqual(key, entry.key)) {
                    return entry;
                }
            }
        }
        return null;
    }

    public void put(K key, V value) {
        var e = _entry(key);
        if (e == null) {
            long h = _hash(key);
            var values = dict.computeIfAbsent(h, k -> new ArrayList<>());
            e = new XEntry<K, V>(key, value);
            values.add(e);
            ++size;
        }
        e.value = value;
    }

    public void remove(K key) {
        long h = _hash(key);
        if (!dict.containsKey(h)) {
            return;
        }
        XEntry<K, V> e = null;
        for (var entry : dict.get(h)) {
            if (feat.isEqual(key, entry.key)) {
                e = entry;
                break;
            }
        }
        if (e != null) {
            dict.get(h).remove(e);
            --size;
        }
    }

    public int size() {
        return size;
    }

}

File: framework/kotlin/src/main/kotlin/soda/kotlin/unittest/validate/XMap.kt
package soda.kotlin.unittest.validate

class XEntry<K, V>(val key: K, var value: V)

class XMap<K, V>(private val feat: ObjectFeature<K>) {

    private val dict = mutableMapOf<Long, MutableList<XEntry<K, V>>>()

    private var size: Int = 0

    private fun _hash(key: K): Long = feat.hash(key)

    private fun _entry(key: K): XEntry<K, V>? {
        return dict.getOrElse(_hash(key)) { mutableListOf() }.find { feat.isEqual(key, it.key) }
    }

    operator fun contains(key: K): Boolean = _entry(key) != null

    operator fun get(key: K): V? = _entry(key)?.value

    fun getOrDefault(key: K, default: V): V {
        return _entry(key)?.value ?: default
    }

    operator fun set(key: K, value: V) {
        val e = _entry(key) ?: run {
            val h = _hash(key)
            val values = dict.getOrPut(h) { mutableListOf() }
            val entry = XEntry(key, value)
            values.add(entry)
            size += 1
            entry
        }
        e.value = value
    }

    operator fun minusAssign(key: K) {
        remove(key)
    }

    fun remove(key: K): V? {
        val h = _hash(key)
        var elem: V? = null
        dict[h]?.let { entries ->
            entries.find { e -> feat.isEqual(key, e.key) }?.let { entry ->
                dict[h]?.remove(entry)
                size -= 1
                elem = entry.value
            }
        }
        return elem
    }
}